src/target/riscv/riscv.c

```diff
@@ -1815,6 +1823,33 @@ int riscv_flush_registers(struct target *target)
 	return ERROR_OK;
 }
 
+static enum target_debug_reason
+derive_debug_reason_without_hitbit(const struct target *target, riscv_reg_t dpc)
+{
+	/* TODO: if we detect that etrigger/itrigger/icount is set, we should
+	 * just report DBG_REASON_UNKNOWN, since we can't disctiguish these
+	 * triggers from BP/WP or from other triggers of such type. However,
+	 * currently this renders existing testsuite as failing. We need to
+	 * fix the testsuite first
+	 */
+	// TODO: the code below does not handle context-aware trigger types
+	for (const struct breakpoint *bp = target->breakpoints; bp; bp = bp->next) {
+		// TODO: investigate if we need to handle bp length
+		if (bp->type == BKPT_HARD && bp->is_set && bp->address == dpc) {
+			// FIXME: bp->linked_brp is uninitialized
+			if (bp->asid) {
+				LOG_TARGET_ERROR(target,
+					"can't derive debug reason for context-aware breakpoint: "
+					"unique_id = %" PRIu32 ", address = %" TARGET_PRIxADDR
+					", asid = %" PRIx32 ", linked = %d",
+					bp->unique_id, bp->address, bp->asid, bp->linked_brp);
+				return DBG_REASON_UNDEFINED;
+			}
+			return DBG_REASON_BREAKPOINT;
+		}
+	}
+	return DBG_REASON_WATCHPOINT;
+}
```

Comment:
@MarekVCodasip I plan to re-write this in a somewhat similar manner (except for calling riscv_hit_watchpoint, more on that later). However, there are few issues.

1. One of them is this part of TODO above:

```
/* TODO: if we detect that etrigger/itrigger/icount is set, we should
	 * just report DBG_REASON_UNKNOWN, since we can't disctiguish these
	 * triggers from BP/WP or from other triggers of such type. However,
	 * currently this renders existing testsuite as failing. We need to
	 * fix the testsuite first
```

That is the testsuite needs to be fixed first. I plan to address this, but not a part of this PR.

2. I'm strongly against any usage of riscv_hit_watchpoint here (or anywhere else). In fact in a subsequent commit I'm going to add a new parameter to get rid of calling this function at all. This function is useless (IMHO) and dangerous .

The reason for such a strong opinion regarding `riscv_hit_watchpoint` is the following.

I. We need to read memory. This is NOT a simple operation. This requires us to figure out the context (like was it user Mode, was there hypervisor enabled, etc). Then (optionally) perform a SW page walk and only then to attempt decode instruction.

II. Decoding instruction by itself is NOT a trivial task. As https://github.com/riscv-collab/riscv-openocd/pull/1029 show it requires significant amount of code that is hard to test. Another problem comes that the list of memory operations depends on extensions supported by target. There is RVV extension, there is also support for stack manipulation extension or whatever abysmal extension will be ratified in future. Unless there is a separate library for this kind of task - I don't think that the respected disassembler belongs to  OpenOCD codebase.

III. Decoding instruction is one thing. We also need to figure out the virtual address. This is NOT a simple task. For some instructions (like indexed loads/stores) from RVV it requires us to know the semantic of instruction and (possibly) additional execution context. That is it requires us to have a mini simulator embedded in OpenOCD. I don't think this is reasonable.

Given points I, II and III I would drop the functionality `riscv_hit_watchpoint` entirely  for future versions and require all targets to support hit bits if they wan't this kind of functionality.

Reply:
> I would drop the functionality `riscv_hit_watchpoint` entirely for future versions and require all targets to support hit bits if they want this kind of functionality.

TLDR: I really dislike `riscv_hit_watchpoint` :)

Reply:
Hello @aap-sc, thank you for explaining your reasoning around `riscv_hit_watchpoint`.

> This function [riscv_hit_watchpoint] is useless (IMHO) and dangerous .

The function allows to inform GDB about watchpoint hit even on targets that do not support the trigger "hit" bits (see the `awatch`, `rwatch` and `watch` stop reasons in the GDB's `T` packet). 

The fact that OpenOCD users went great lengths to implement this mechanism shows that it was important to them. The mechanism need not be perfect to help users. If there are cases that we do not cover, like V extension, it would be good to warn the user.

> We need to read memory. This is NOT a simple operation. 

Reading memory -- one of the fundamental debugger operations -- should work reliably. If there are any concerns about certain cases not handled well, then that should be addressed as an independent issue (irrespective of riscv_hit_watchpoint()).

> Decoding instruction by itself is NOT a trivial task. As https://github.com/riscv-collab/riscv-openocd/pull/1029 show it requires significant amount of code that is hard to test. 

The code for decoding would be good candidate for unit-testing. I am considering whether to add unit-test capability to OpenOCD. 

> Another problem comes that the list of memory operations depends on extensions supported by target. (...)

> We also need to figure out the virtual address. This is NOT a simple task. For some instructions (like indexed loads/stores from RVV) it requires us to know the semantic of instruction and (possibly) additional execution context. (...)

The current version of the code (including https://github.com/riscv-collab/riscv-openocd/pull/1029) covers `I` and `C` extensions, which is a good baseline that will probably help significant amount of users. 

If an unsupported case is detected (like V extension), a warning can be issued. 

Alternatively, the user can warn the user anytime we detect that the user works with a core that does not support the "hit" bits.

---

TL,DR: I agree about the limitations of `riscv_hit_watchpoint()`, but still believe that the current version helps users. We should not strive to cover all possible cases. Instead, we can stick to the basics and warn the user about the limitations of the current algorithm.

Reply:
> The fact that OpenOCD users went great lengths to implement this mechanism shows that it was important to them. 

Well, yeah. Probably there are cases when it is useful. Probably, my opinion is way too strong :).

> The mechanism need not be perfect to help users.

Yes, there are cases when it is terrible (see below). So I want to have an option to avoid these memory reads to reduce evasiveness. For example if we debug ONLY with OpenOCD without involvement of GDB we don't really need this functionality. 

> Reading memory -- one of the fundamental debugger operations -- should work reliably. 

The thing is that it **may not** not work reliably in all cases. For example: I constantly see designs that is capable to hang if there is something wrong with the system bus. For example, the underlying device does not send a response. We must not forget that people rarely use debugging with OpenOCD when everything is fine. One uses a debugger when something is (terribly) wrong. 

Reading memory is WAY too intrusive for SOME scenarios.

Reply:
I've created https://github.com/riscv-collab/riscv-openocd/issues/1076 (I plan to start working on that as soon as this patch is landed).

Reply:
Thank you :+1: